tensorflow/core/kernels/reshape_util_gpu.cu.cc

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+
+#define EIGEN_USE_GPU
+
+#include "tensorflow/core/kernels/reshape_util.h"
+#include "tensorflow/core/util/gpu_kernel_helper.h"
+
+namespace tensorflow {
+
+using GPUDevice = Eigen::GpuDevice;
+
+namespace {
+
+template <typename Tindex>
+__global__ void ReshapeSparseTensorKernel(
+    const Tindex nnz, const Tindex input_rank, const Tindex output_rank,
+    const Tindex* __restrict__ input_shape,
+    const Tindex* __restrict__ output_shape,
+    const Tindex* __restrict__ input_indices,
+    Tindex* __restrict__ output_indices) {
+  GPU_1D_KERNEL_LOOP(sparse_index, nnz) {
+    const Tindex* input_index = &input_indices[sparse_index * input_rank];
+    Tindex* output_index = &output_indices[sparse_index * output_rank];
+    Tindex dense_index = 0;
+    // Flatten input index from slowest- to fastest-changing dimension.
+    for (int i = 0; i < input_rank; ++i) {
+      dense_index = dense_index * input_shape[i] + input_index[i];
+    }
+    // Compute output index from fastest- to slowest-changing dimension.
+    for (int i = output_rank; i-- > 0;) {
+      Tindex output_size = output_shape[i];
+      output_index[i] = dense_index % output_size;
+      dense_index /= output_size;
+    }
```

Comment:
Do you need to care about integer overflow in case the indices are in int32?  Maybe always use int64 for `dense_index`?